File: Benchmarking/30 Euler/euler4/euler4.jl
#=
A palindromic number reads the same both ways. The largest palindrome made
from the product of two 2-digit numbers is 9009 = 91 × 99.
Find the largest palindrome made from the product of two 3-digit numbers.
=#

#Answer: 906609

#Terminal: Julia euler4.jl 100 999
#Interval for calculating palindrome: 100-999 finds largest 3-digit palindrome.
min = parse(Int, ARGS[1])
max = parse(Int, ARGS[2])
function digitsInArray(n)
  arr = Int[]
  while n > 0
      d = round(Int, floor(n / 10))
      k = n - d * 10
      n = d
      push!(arr, k)
  end
  b = reverse(arr)
  return b
end

function isPalindrome(array)
  i = 1
  j = length(array)
  isPal = true
    while i <= j
      if array[i] == array[j]
        i += 1
        j -= 1
      else
        isPal = false
        return isPal
      end
    end
  return isPal
end

function checkNumber(min, max)
  maxNum = 0
  found = false
  cycleDone = false
  while !found
    for x in max:-1:min
      for y in max:-1:min
        k = x * y
        result = isPalindrome(digitsInArray(k))
        #println(k)
        if result
          if k > maxNum
            maxNum = k
            #println(max)
            #resultArray = [x, y]
            found = true
          end
        end
      end
    end
    cycleDone = true
    if cycleDone && !found
      break
    end
  end
  return maxNum
end

checkNumber(min, max)

File: Benchmarking/stats/euler4/euler4.cpp
//
//  main.cpp
//  Euler4
//
//  Created by Sonni on 19/05/16.
//  Copyright © 2016 Sonni Jensen. All rights reserved.
//

#include <iostream>
#include <vector>
#include <math.h>
#include <string>
#include <algorithm>

std::vector<unsigned int> digitsInArray(unsigned int n)
{
    std::vector<unsigned int> array;
    unsigned int d;
    unsigned int k;
    while (n > 0)
    {
        d = floor(n / 10);
        k = n - d * 10;
        n = d;
        array.push_back(k);
    }
    std::reverse(array.begin(), array.end());
    return array;
}

bool isPalindrome(std::vector<unsigned int> array)
{
     unsigned int i = 0;
     unsigned long j =  array.size()-1;
    bool isPal = true;
    
    while (i <= j)
    {
        if (array[i] == array[j])
        {
            i++;
            j--;
        }
        else
        {
            isPal = false;
            break;
        }
    }
    
    return isPal;
}

 unsigned int checkNumber( unsigned int min,  unsigned int max)
{
    //unsigned int resultArray[2];
     unsigned int maxNum = 0;
    bool found = false;
    bool cycleDone = false;
    
    while (!found)
    {
        for ( unsigned int x = max; x >= min; x--)
        {
            for ( unsigned int y = max; y >= min; y--)
            {
                 unsigned int k = x * y;
                bool result = isPalindrome(digitsInArray(k));
                if (result)
                {
                    if (k > maxNum)
                    {
                        maxNum = k;
                        
                        //resultArray[0] = x;
                        //resultArray[1] = y;
                        found = true;
                    }
                }
            }
        }
        cycleDone = true;
        if (!found)
            break;
    }
    return maxNum;
}

int main(int argc, const char * argv[]) {
    std::cout << checkNumber(atoi(argv[1]), atoi(argv[2])) << std::endl;
    return 0;
}
